app/pages/settings/email.jsx

```diff
@@ -0,0 +1,289 @@
+import React from 'react';
+import apiClient from 'panoptes-client/lib/api-client';
+import talkClient from 'panoptes-client/lib/talk-client';
+import AutoSave from '../../components/auto-save';
+import handleInputChange from '../../lib/handle-input-change';
+
+class EmailSettingsPage extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      page: 1,
+      projects: [],
+      projectPreferences: [],
+      talkPreferences: []
+    };
+    this.handleProjectPreferenceChange = this.handleProjectPreferenceChange.bind(this);
+    this.handleTalkPreferenceChange = this.handleTalkPreferenceChange.bind(this);
+    this.talkPreferenceOption = this.talkPreferenceOption.bind(this);
+    this.getProjectForPreferences(props.user);
+    this.getTalkPreferences();
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    if (!this.state.page === prevState.page) {
+      this.getProjectForPreferences();
+    }
+  }
+
+  getTalkPreferences() {
+    talkClient.type('subscription_preferences').get()
+    .then((preferences) => {
+      this.setState({
+        talkPreferences: preferences
+      });
+    });
+  }
+
+  getProjectForPreferences(user) {
+    return user.get('project_preferences', {
+      page: this.state.page,
+      sort: 'display_name'
+    })
+    .then((preferences) => {
+      if (preferences.length > 0) {
+        const projectIDs = preferences.map(pref => pref.links.project);
+        apiClient.type('projects').get({
+          id: projectIDs
+        })
+        .then((projects) => {
+          this.setState({
+            meta: preferences[0].getMeta(),
+            projectPreferences: preferences,
+            projects
+          });
+        });
+      }
+    })
+    .catch((error) => {
+      console.log('Something went wrong. Error: ', error);
+    });
+  }
+
+  handleProjectPreferenceChange(preference, event) {
+    preference.update({
+      email_communication: !!event.target.checked
+    })
+    .save()
+    .then(updatedPref => this.setState(updatedPref));
```

Comment:
This overwrites component state with a preference resource, leading to the creation of `state.id`, `state._type` etc. You can check this in the React inspector in dev tools.

I think what you mean to update here is `state.projectPreferences`.

Reply:
Ah my bad, thanks. I'm struggling to merge the updated resource into the existing array. How should I do that?

Reply:
`preference` is a reference to an item in the `projectPreferences` array, not a copy, so you're already updating the array. 

If you want to make it this function more explicit, you could pass the array index instead of the preference, then `projectPreferences[index].update(changes).save()` and `setState({ projectPreferences })`.

Reply:
Oh wait, you want to store the updated preference from the API as `projectPreferences[index]` before you set state. The key is still to pass the array index, rather than the preference itself. 

Or you could get the index from `projectPreferences.indexOf(preference)`, if you leave the handler as it is and pass the preference in.

Reply:
Thanks, just pushed latest fixes.